### ISSUE TITLE
Please support Results<T> as return type of controllers

### ISSUE THREAD

**Issue by aureole82** [domaindrivendev/Swashbuckle.AspNetCore#2595]
Introducing .NET 7 `Results<T>` (https://learn.microsoft.com/en-us/aspnet/core/web-api/action-return-types#resultt-type) Microsoft promised that
> All the [[ProducesResponseType]](https://learn.microsoft.com/en-us/dotnet/api/microsoft.aspnetcore.mvc.producesresponsetypeattribute) attribute's can be excluded, since the HttpResult implementation contributes automatically to the endpoint metadata.

But if I write this:
```c#
/// <summary> Request the forecast of a specific day. </summary>
/// <param name="date" example="2023-01-24">Date of the requested forecast. </param>
[HttpGet("{date}", Name = "GetWeatherForecastByDate")]
public Results<Ok<WeatherForecast>, NotFound> Get(DateOnly date)
{
    var forecast = _weatherForecasts.FirstOrDefault(forecast => forecast.Date == date);
    return forecast != null ? TypedResults.Ok(forecast) : TypedResults.NotFound();
}
```
The generated `swagger.json` only contains that:
```json
"responses": {
  "200": {
    "description": "Success"
  }
}
```
<hr />

In order to get the correct schema I still need to add the following `ProducesResponseTypeAttribute`s :-(
```c#
[ProducesResponseType(typeof(WeatherForecast), (int)HttpStatusCode.OK)]
[ProducesResponseType((int)HttpStatusCode.NotFound)]
```

**Comment by hjrb** [comment]
I also need this featue

**Comment by opflucker** [comment]
I also need this featue. Thank for your support 👍👍🙏

**Comment by vernou** [comment]
It's a know issue. You can read this GitHub issue to full detail :\
[TypedResults metadata are not inferred for API Controllers](https://github.com/dotnet/aspnetcore/issues/44988)

In summary, metadata aren't correctly inferred to action that return type HttpResults... but sadly, Swashbuckle need this metadata to generate the OpenApi schema.

---

While waiting for a resolution, I do a operation filter to generate OpenApi response from HttpResults type and I wrapped it in a NuGet package :\
[Vernou.Swashbuckle.HttpResultsAdapter](https://www.nuget.org/packages/Vernou.Swashbuckle.HttpResultsAdapter)

You can add the package :
```sh
dotnet add package Vernou.Swashbuckle.HttpResultsAdapter
```

So you can register the filter :
```csharp
var builder = WebApplication.CreateBuilder(args);
...
builder.Services.AddSwaggerGen(options =>
{
    ...
    options.OperationFilter<Vernou.Swashbuckle.HttpResultsAdapter.HttpResultsOperationFilter>();
});
```

If you prefer, you can just copy/paste the filter file from the package repository :\
[HttpResultsOperationFilter.cs](https://github.com/vernou/Vernou.Swashbuckle.HttpResultsAdapter/blob/main/src/Vernou.Swashbuckle.HttpResultsAdapter/HttpResultsOperationFilter.cs)

**Comment by hjrb** [comment]
Hi

> It's a know issue. You can read this GitHub issue to full detail : [TypedResults metadata are not inferred for API Controllers](https://github.com/dotnet/aspnetcore/issues/44988)
> 
> In summary, metadata aren't correctly inferred to action that return type HttpResults... but sadly, Swashbuckle need this metadata to generate the OpenApi schema.
> 
> While waiting for a resolution, I do a operation filter to generate OpenApi response from HttpResults type and I wrapped it in a NuGet package : [Vernou.Swashbuckle.HttpResultsAdapter](https://www.nuget.org/packages/Vernou.Swashbuckle.HttpResultsAdapter)
> 
> You can add the package :
> 
> ```shell
> dotnet add package Vernou.Swashbuckle.HttpResultsAdapter
> ```
> 
> So you can register the filter :
> 
> ```cs
> var builder = WebApplication.CreateBuilder(args);
> ...
> builder.Services.AddSwaggerGen(options =>
> {
>     ...
>     options.OperationFilter<Vernou.Swashbuckle.HttpResultsAdapter.HttpResultsOperationFilter>();
> });
> ```
> 
> If you prefer, you can just copy/paste the filter file from the package repository : [HttpResultsOperationFilter.cs](https://github.com/vernou/Vernou.Swashbuckle.HttpResultsAdapter/blob/main/src/Vernou.Swashbuckle.HttpResultsAdapter/HttpResultsOperationFilter.cs)

Hi, I gave your package a try. But there is one issue: most of my controller methods are async. Hence they return a Task<HttpResults>. Your code does not account for this. I'll have a deeper look into your code and send you an suggestion how to hande async

**Comment by vernou** [comment]
@hjrb, thank for you return.

The support to async action was added in the version [1.0.1](https://www.nuget.org/packages/Vernou.Swashbuckle.HttpResultsAdapter/1.0.1).

**Comment by hjrb** [comment]
excellent

**Comment by Havunen** [comment]
FYI. this is supported in DotSwashbuckle v3.0.8+

**Comment by jgarciadelanoceda** [comment]
@aureole82 

Could you please verify if this is still open?

I have verified using the latest version of SwashBuckle and it seems to work.

![image](https://github.com/user-attachments/assets/a2e4c9cd-e82e-4ed0-bac9-8af58b6ad678)


The result OpenApi document is OK having all the info.

**Comment by vernou** [comment]
From the documentation [Controller action return types in ASP.NET Core web API](https://learn.microsoft.com/en-us/aspnet/core/web-api/action-return-types?view=aspnetcore-8.0#resultstresult1-tresultn-type) :
> ### Results<TResult1, TResultN> type
> The static [TypedResults](https://learn.microsoft.com/en-us/dotnet/api/microsoft.aspnetcore.http.typedresults) class returns the concrete IResult implementation that allows using IResult as return type. The usage of the concrete IResult implementation offers the following benefit over the [IResult type](https://learn.microsoft.com/en-us/aspnet/core/web-api/action-return-types?view=aspnetcore-8.0#iresult-type):
> ...

So it sound a legit use case.

**Comment by jgarciadelanoceda** [comment]
Yeah, I showed it, but I will in any case create an issue in open-api dotnet and wait for a new release

**Comment by aureole82** [comment]
> FYI. this is supported in DotSwashbuckle v3.0.8+

I have just tested it with .NET 8 and [Swashbuckle.AspNetCore v6.6.2](https://www.nuget.org/packages/Swashbuckle.AspNetCore/6.6.2) and it still doesn't produce any response type:
```json
"/WeatherForecast/{date}": {
  "get": {
    "tags": [
      "WeatherForecast"
    ],
    "operationId": "GetWeatherForecastByDate",
    "parameters": [
      {
        "name": "date",
        "in": "path",
        "required": true,
        "schema": {
          "type": "string",
          "format": "date"
        }
      }
    ],
    "responses": {
      "200": {
        "description": "OK"
      }
    }
  }
}
```

**Comment by jgarciadelanoceda** [comment]
Yeah, I can indeed see the commit that was done in DotnetSwashbuckle, but as is not available either in new openApi package of MS, I have notified them

**Comment by jgarciadelanoceda** [comment]
@martincostello I am thinking about applying the same fix that was done in DotSwashbuckle. I created a PR to fix this on aspnetCore itself but I am not sure if they are going to backport it.
dotnet/aspnetcore#57464

**Comment by martincostello** [comment]
I think it's fair game for any changes in DotSwashbuckle to fix user's issues to be ported back here.

**Comment by VAllens** [comment]
> > FYI. this is supported in DotSwashbuckle v3.0.8+
> 
> I have just tested it with .NET 8 and [Swashbuckle.AspNetCore v6.6.2](https://www.nuget.org/packages/Swashbuckle.AspNetCore/6.6.2) and it still doesn't produce any response type:
> 
> "/WeatherForecast/{date}": {
>   "get": {
>     "tags": [
>       "WeatherForecast"
>     ],
>     "operationId": "GetWeatherForecastByDate",
>     "parameters": [
>       {
>         "name": "date",
>         "in": "path",
>         "required": true,
>         "schema": {
>           "type": "string",
>           "format": "date"
>         }
>       }
>     ],
>     "responses": {
>       "200": {
>         "description": "OK"
>       }
>     }
>   }
> }

I'm use [Swashbuckle.AspNetCore v8.1.3](https://www.nuget.org/packages/Swashbuckle.AspNetCore/8.1.3) in the Minimal APIs with .NET 9.0.
I'm having the same problem, any progress?

**Comment by jgarciadelanoceda** [comment]
This is solved in aspnetcore itself see dotnet/aspnetcore#60096

**Comment by martincostello** [comment]
We should test that Swashbuckle.AspNetCore actually works with it correctly in #3283.

**Comment by martincostello** [comment]
I'm going to close this out - we've validated it works for .NET 10.
